tests/unit/test_stubs.py

```diff
@@ -21,3 +23,52 @@ def testing_connect(*args):
             vcr_connection.cassette = Cassette("test", record_mode=mode.ALL)
             vcr_connection.real_connection.connect()
             assert vcr_connection.real_connection.sock is not None
+
+    def test_body_consumed_once_stream(self, tmpdir, httpbin):
+        self._test_body_consumed_once(
+            tmpdir,
+            httpbin,
+            "body_consumed_once_stream.yml",
+            BytesIO(b"1234567890"),
+            BytesIO(b"9876543210"),
+            BytesIO(b"9876543210"),
+        )
+
+    def test_body_consumed_once_iterator(self, tmpdir, httpbin):
+        self._test_body_consumed_once(
+            tmpdir,
+            httpbin,
+            "body_consumed_once_iterator.yml",
+            iter([b"1234567890"]),
+            iter([b"9876543210"]),
+            iter([b"9876543210"]),
+        )
+
+    def _test_body_consumed_once(
+        self,
+        tmpdir,
+        httpbin,
+        testfile,
+        data1,
+        data2,
+        data3,
+    ):
+        testpath = str(tmpdir.join(testfile))
```

Comment:
Any particular reason why we want control over the precise filename here? If not, `with NamedTemporaryFile(...)` would work: 
- `suffix=".yml"` allows to control the file extension if needed
- `dir=tmpdir` allows using the directory we have been given.

It would resolve one no-longer-necessary parameter. I think it could help simplicity.

Reply:
I had to use `TemporaryDirectory` instead, because vcrpy fails when cassette is an empty file.

Reply:
@sathieu interesting! I got curious, played with it and can confirm that the file cannot exits already when we start. Here's what I did to make it work still:

```diff
# git --no-pager diff
diff --git a/tests/unit/test_stubs.py b/tests/unit/test_stubs.py
index c177061..8373923 100644
--- a/tests/unit/test_stubs.py
+++ b/tests/unit/test_stubs.py
@@ -1,7 +1,7 @@
 import contextlib
 import http.client as httplib
 from io import BytesIO
-from tempfile import TemporaryDirectory
+from tempfile import NamedTemporaryFile
 from unittest import mock
 
 from pytest import mark
@@ -52,8 +52,12 @@ class TestVCRConnection:
         data2,
         data3,
     ):
-        with TemporaryDirectory(dir=tmpdir) as cassete_dir:
-            testpath = str(cassete_dir.join("cass.yml"))
+        with NamedTemporaryFile(dir=tmpdir, suffix='.yml') as f:
+            testpath = f.name
+            # NOTE: ``use_cassette`` is not okay with the file existing
+            #       already.  So we using ``.close()`` to not only
+            #       close but also delete the empty file, before we start.
+            f.close()
             host, port = httpbin.host, httpbin.port
             match_on = ["method", "uri", "body"]
             with use_cassette(testpath, match_on=match_on):
```

I'm not saying that's better, maybe it's not :shrug: How about adding a sentence of comment why `TemporaryDirectory` is prefered over `NamedTemporaryFile` in our case? Something like that would maybe be best?

Reply:
Yes, changed to your suggestion. thanks!